pkg/volume/azure_file/azure_file.go

```diff
@@ -0,0 +1,237 @@
+	return plugin.newBuilderInternal(spec, pod, &azureSvc{}, plugin.host.GetMounter())
+}
+
+func (plugin *azureFilePlugin) newBuilderInternal(spec *volume.Spec, pod *api.Pod, util azureUtil, mounter mount.Interface) (volume.Builder, error) {
+	var source *api.AzureFileVolumeSource
+	var readOnly bool
+	if spec.Volume != nil && spec.Volume.AzureFile != nil {
+		source = spec.Volume.AzureFile
+		readOnly = spec.Volume.AzureFile.ReadOnly
+	} else {
+		source = spec.PersistentVolume.Spec.AzureFile
+		readOnly = spec.ReadOnly
+	}
+	return &azureFileBuilder{
+		azureFile: &azureFile{
+			volName: spec.Name(),
+			mounter: mounter,
+			pod:     pod,
+			plugin:  plugin,
+		},
+		util:        util,
+		accountName: source.AccountName,
+		accountKey:  source.AccountKey,
+		shareName:   source.ShareName,
+		readOnly:    readOnly,
+	}, nil
+}
+
+func (plugin *azureFilePlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return plugin.newCleanerInternal(volName, podUID, plugin.host.GetMounter())
+}
+
+func (plugin *azureFilePlugin) newCleanerInternal(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	return &azureFileCleaner{&azureFile{
+		volName: volName,
+		mounter: mounter,
+		pod:     &api.Pod{ObjectMeta: api.ObjectMeta{UID: podUID}},
+		plugin:  plugin,
+	}}, nil
+}
+
+// azureFile volumes represent mount of an AzureFile share.
+type azureFile struct {
+	volName string
+	pod     *api.Pod
+	mounter mount.Interface
+	plugin  *azureFilePlugin
+}
+
+func (azureFileVolume *azureFile) GetPath() string {
+	name := azureFilePluginName
+	return azureFileVolume.plugin.host.GetPodVolumeDir(azureFileVolume.pod.UID, util.EscapeQualifiedNameForDisk(name), azureFileVolume.volName)
+}
+
+type azureFileBuilder struct {
+	*azureFile
+	util        azureUtil
+	accountName string
+	accountKey  string
+	shareName   string
+	readOnly    bool
+}
+
+var _ volume.Builder = &azureFileBuilder{}
+
+func (_ *azureFileBuilder) SupportsOwnershipManagement() bool {
+	return false
+}
+
+// SetUp attaches the disk and bind mounts to the volume path.
+func (b *azureFileBuilder) SetUp() error {
+	return b.SetUpAt(b.GetPath())
+}
+
+func (b *azureFileBuilder) SetUpAt(dir string) error {
+	notMnt, err := b.mounter.IsLikelyNotMountPoint(dir)
+	glog.V(4).Infof("AzureFile mount set up: %s %v %v", dir, !notMnt, err)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if !notMnt {
+		return nil
+	}
+	if err = b.util.SetupAzureFileSvc(b.accountName, b.accountKey, b.shareName); err != nil {
+		return err
+	}
+	os.MkdirAll(dir, 0750)
+	source := fmt.Sprintf("//%s.file.core.windows.net/%s", b.accountName, b.shareName)
+	// parameters suggested by https://azure.microsoft.com/en-us/documentation/articles/storage-how-to-use-files-linux/
+	options := []string{fmt.Sprintf("vers=3.0,username=%s,password=%s,dir_mode=0777,file_mode=0777", b.accountName, b.accountKey)}
```

Comment:
Hmm if accountKey is your password then revealing it in the pod definition might be problematic.